### PR TITLE
fix duplicate snapshots when using multiple labelSelector schedules

### LINKED ISSUE #1063: cmd/restic doesn't respect labelSelector and duplicates backups
### Description

k8up's labelSelector feature allows targeting specific preBackupPods and PVCs.

I've observed some behavior causing duplicate restic snapshots, uploading them to unrelated repositories.

This happens when multiple preBackupPods are spawned in parallel. The restic Job spec is missing the TARGET_PODS env, needed by: https://github.com/k8up-io/k8up/blob/master/cmd/restic/main.go#L110

This bug has been originally introduced: https://github.com/k8up-io/k8up/pull/1045/

I will submit a fix for it - already looked into it quite a bit. I aim to have the PR out this week.  

### Additional Context

_No response_

### Logs

```shell

```

### Expected Behavior

When defining multiple schedules, targeting multiple buckets, only the intended backups get uploaded. 

### Steps To Reproduce

1. define 2 different schedules, each targeting different datastores with labelSelectors and different buckets
2. adjust them to run at the exact same time
3. The data will be uploaded to both buckets, coming from both schedules
### Version of K8up

2.12.0

### Version of Kubernetes

1.22

### Distribution of Kubernetes

all

### PR DESCRIPTION
**Pull request by danielpodwysocki**
## Summary

fixes #1063

This fixes duplicate snapshots being created in a following scenario.

Assuming you have backup targets:

preBackupPod -> labeled "target: mytarget1"
preBackupPod-> labeled "target: mytarget2"

And following backup schedules:

schedule1 -> targets mytarget1 via labelSelector
schedule2 -> targets mytarget2 via labelSelector

If both schedules run at the same time, the cmd/restic module will create snapshots for both - TARGET_PODS is not passed and all viable annotated pods are included for both.

This can cause compliance problems - the schedules could be targeting different buckets. That means data not meant to end up in one of them could get uploaded there.

labelSelector references for context:
- https://github.com/k8up-io/k8up/pull/1045
- https://docs.k8up.io/k8up/2.12/how-tos/backup.html#_target_specific_pvcs_or_prebackuppods

### For Code changes

- [x] Categorize the PR by setting a good title and adding one of the labels:
  `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
  as they show up in the changelog
- [x] PR contains the label `area:operator`
- [ x ] Commits are [signed off](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits)
- [ x ] Link this PR to related issues
- [ x ] I have not made _any_ changes in the `charts/` directory.

### For Helm Chart changes

- [ ] Categorize the PR by setting a good title and adding one of the labels:
  `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
  as they show up in the changelog
- [ ] PR contains the label `area:chart`
- [ ] PR contains the chart label, e.g. `chart:k8up`
- [ ] Commits are [signed off](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits)
- [ ] Variables are documented in the values.yaml using the format required by [Helm-Docs](https://github.com/norwoodj/helm-docs#valuesyaml-metadata).
- [ ] Chart Version bumped if immediate release after merging is planned
- [ ] I have run `make chart-docs`
- [ ] Link this PR to related code release or other issues.

### CODE REVIEW

File: operator/backupcontroller/executor.go [hunk 1]
```diff
@@ -232,17 +231,24 @@ func (b *BackupExecutor) startBackup(ctx context.Context) error {
 	}
 
 	log := controllerruntime.LoggerFrom(ctx)
-	podLister := kubernetes.NewPodLister(ctx, b.Client, cfg.Config.BackupCommandAnnotation, "", "", b.backup.Namespace, nil, false, log)
-	backupPods, err := podLister.ListPods()
+	// validating that the correct backup annotations exist is already handled in the restic module
+	// we only need to pass it a list of pods it should go through
+
+	backupPods := &corev1.PodList{}
+	err = b.fetchCandidatePods(ctx, backupPods)
```
[the remark this answers is not in the document]
Comment (Kidswiss) [reply operator/backupcontroller/executor.go:238]: That's tricky. Most of the filtering concerning pre-backup pods is done `cmd/restic` if I'm not mistaken.

I'd probably just keep it as is for now. I can't think of any down- or upside to either approach. Except that one might be a bit simpler to implement than the other.

**Comment by Kidswiss** [comment]
Thanks for the PR! That's quite the edge case.

So I assume it really only happens if multiple backups run at exactly the same time? 

> Looking at the tests - my bad, I ran a different test target before pushing it out.

No problem, just ping again once you've fixed them and then I will trigger the actions again.

**Comment by Kidswiss** [comment]
Hi @danielpodwysocki 

Were you able to get back to this?

**Comment by danielpodwysocki** [comment]
Hey @Kidswiss ,

Not yet - seems this has fallen through the cracks, my bad! 🙈 
I'll cut out some time for it and have a look over the weekend.

In case you need it sooner and you'd like to grab what's already there and adjust it, feel free to as well.

**Comment by Kidswiss** [comment]
@danielpodwysocki 

No worries. :) If you can do it over the weekend it should be fine. Then I can do a new release next week.

**Comment by danielpodwysocki** [comment]
@Kidswiss I've fixed up the tests - the failing ones were in `make integration-test`.

The backup already expects the filtering from `cmd/restic` to happen by the time the TARGET_PODS are passed - I've done that and both e2e&integration tests pass now.

**Comment by Kidswiss** [comment]
@danielpodwysocki Thanks! Merging now.
